[Yong Chang Sheng] iP

rycs2812 opened the pull request:
## Duke
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

Duke frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/rycs2812/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

##
If you're a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

rycs2812 committed:
Implement greet, echo, exit

rycs2812 committed:
Implement adding to, displaying list

rycs2812 committed:
Add task marking, unmarking

rycs2812 committed:
Add support for todo, event, deadline tasks

rycs2812 committed:
Implement automated testing

rycs2812 committed:
Add support for deleting tasks

rycs2812 committed:
Refactor code to use enums, update expected.txt for UI testing

rycs2812 committed:
Add automatic saving of tasks, data/nuke.txt to .gitignore

rycs2812 committed:
Refactor code, dates represented with LocalDate instead of String class

rycs2812 committed:
Merge branch 'branch-Level-7'

rycs2812 committed:
Merge branch 'branch-Level-8'

rycs2812 committed:
Fix bugs after branch merge

rycs2812 committed:
Refactor code to extract out closely related code as classes

rycs2812 committed:
Organise classes into java packages

rycs2812 committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2223S1/ip

rycs2812 committed:
Add JUnit tests

rycs2812 committed:
Add JavaDoc comments to the code

rycs2812 committed:
Tweak code to comply with module coding standard

rycs2812 committed:
Implement find command

rycs2812 committed:
Merge branch 'branch-A-JavaDoc'

rycs2812 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Command.java
#	src/main/java/duke/Storage.java

rycs2812 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Command.java
#	src/main/java/duke/TaskList.java

rycs2812 committed:
Fix indentation of switch-case statement in Command.java

rycs2812 committed:
Switch access modifier of Command.java methods to private

EiffelLKF commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,252 @@
+        }
+    }
+
+    private boolean isValidDate(String date) {
+        try {
+            LocalDate d = LocalDate.parse(date, DateTimeFormatter.ofPattern(INPUT_DATE_FORMAT));
+            return true;
+        } catch (DateTimeParseException e) {
+            return false;
+        }
+    }
```
You can consider adding header comments to all your methods to help developers better understand the code

EiffelLKF commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,252 @@
+        }
+    }
+
+    private boolean isCorrectUsage(String args, Duke duke) {
```
Good naming choice for your boolean methods, very 'boolean-sounding' ! 😄

EiffelLKF commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,252 @@
+            }
+            return true;
+
+        case "mark":
```
You might need to add an explicit "// fallthrough" comment when you are using the same code for multiple cases

EiffelLKF commented on src/main/java/duke/UI.java:
```diff
@@ -0,0 +1,92 @@
+        }
+    }
+
+    public void print(String s) {
+        System.out.println(LINE_BREAK);
+        System.out.println(s);
+        System.out.println(LINE_BREAK);
+        System.out.println();
+    }
```
Similar issue about adding header comments

EiffelLKF commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,120 @@
+            }
+            String task = "";
+            for (int i = 3; i < temp.length - 1; ++i) {
+                if (i < temp.length - 2) {
+                    task += temp[i];
+                    task += "[";
+                } else {
+                    task += temp[i].substring(0, temp[i].length() - 1);
+                }
```
This might be considered deep nesting, is there another way you can refactor this?

EiffelLKF commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,120 @@
+            String nextTask = s.nextLine();
+            String[] temp = nextTask.split("\\[");
+            Character taskType = temp[1].charAt(0);
+            Boolean marked = temp[2].charAt(0) == 'X';
```
Perhaps you can consider changing this to have a more 'boolean-sounding' name convention to be more explicit

EiffelLKF commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,120 @@
+        ArrayList<Task> tasks = new ArrayList<>();
+        Scanner s = new Scanner(f);
+        while(s.hasNext()) {
+            String nextTask = s.nextLine();
+            String[] temp = nextTask.split("\\[");
+            Character taskType = temp[1].charAt(0);
+            Boolean marked = temp[2].charAt(0) == 'X';
+            String dateTemp = temp[temp.length - 1];
+            dateTemp = dateTemp.substring(0, dateTemp.length() - 1);
+            LocalDate date = LocalDate.now();
```
This portion of the code can be a bit hard to read and understand, it would be nice to include some logical spaces and comments to help the read understand it more easily

EiffelLKF commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,252 @@
+            }
+            break;
+
+        case "mark":
```
You might need an explicit "// fallthrough" comment here

EiffelLKF commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,252 @@
+                ui.printError("Please specify at least one keyword.\n\n" + usage);
+                return false;
+            } else if (args.charAt(0) != '"' || args.charAt(args.length() - 1) != '"') {
+                if (args.split(" ").length > 1) {
+                    ui.printError("'find' only expects 1 argument.\n\n If there are multiple "
+                            + "keywords, please enclose them in quotation marks (\"\").");
+                    return false;
+                }
```
This might be considered deep nesting, is there any way for you to refactor this?

EiffelLKF commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,252 @@
+                    return false;
+                }
+            } else {
+                if (args.length() == 2 || args.substring(1, args.length() - 1).isBlank()) {
```
This condition is a bit complex, it might help to abstract some of this information into a method or evaluate these statements in separate intemediate steps?

EiffelLKF commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,252 @@
+            } else {
+                try {
+                    int index = Integer.parseInt(args) - 1;
+                    if (index < 0 || index >= taskList.getCount()) {
+                        ui.printError("Please specify a valid task number.\n"
+                                + "There are " + taskList.getCount()
+                                + " task(s) in the list.\n\n" + usage);
+                        return false;
+                    }
```
Similar thoughts on deep nesting

EiffelLKF commented on src/main/java/duke/UI.java:
```diff
@@ -0,0 +1,92 @@
+/**
+ * An input/output handler that deals with interaction with the user.
+ */
+public class UI {
```
You might want to change this class name is "Ui" to follow coding standard for acronym names

EiffelLKF reviewed (commented):
Hey! Great work on the application! I like the way you structured your code logically, but it might help to add some comments and space breaks to improve readability. Overall, it looks really good and with some small tweaks, I'm sure it will be a great project! 😃

yuxuanleong commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,73 @@
+        return tasks.size();
+    }
+
+    public ArrayList<Task> getTasks() {
+        return tasks;
+    }
```
I'm slightly concern about the public `getTasks()` method as anyone can access this and they can modify your array. Alternatively, I was thinking this method can be a `getTask(int index)` method that returns `Task` at a specific index instead of `ArrayList<Task>`. In the rest of the files, the task list can be passed in as parameter and in the body you can `getTask(int index)` and then continue with the logic.

yuxuanleong commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,45 @@
+import java.io.FileNotFoundException;
+
+/**
+ * A Duke bot.
```
Would it be better if the description is slightly more detailed?

yuxuanleong commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,252 @@
+        }
+        switch(name) {
+        case "delete":
+            int index = Integer.parseInt(args) - 1;
```
What will happen if Integer is not able to parse the int input by the user?

yuxuanleong commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,252 @@
+
+    private String INPUT_DATE_FORMAT = "dd-MM-yyyy";
+
+    private final String name;
+    Command(String name) {
+        this.name = name;
+    }
```
It would be good to add a line break.

yuxuanleong reviewed (commented):
I really like the structure of the classes and methods. Some of the things to note are formatting layouts (eg. import statements, adding line breaks where necessary) as well as some missing JavaDocs (for public methods).

rycs2812 committed:
Configure checkstyle and fix styling issues

rycs2812 committed:
Merge branch 'branch-A-CheckStyle'

rycs2812 committed:
Add basic GUI to Duke

rycs2812 committed:
Merge branch 'branch-Level-10'

rycs2812 committed:
Implement Duke functionality to GUI

rycs2812 committed:
Remove text UI automated testing

rycs2812 committed:
Add assertions

rycs2812 committed:
Refactor code to improve code quality

Modified program so that the execution of commands (e.g. adding task to
the task list and also writing to storage) is handled by Duke.java
instead of Command.java.

Also shortened methods in Command.java and Storage.java to improve
readability, and changed names of Boolean variables in accordance with
the CS2103T naming conventions.

rycs2812 committed:
Merge pull request #2 from rycs2812/branch-A-Assertions

Add assertions

rycs2812 committed:
Merge branch 'master' into branch-A-CodeQuality

rycs2812 committed:
Merge pull request #3 from rycs2812/branch-A-CodeQuality

Refactor code to improve code quality

rycs2812 committed:
Implement recurring deadlines

rycs2812 committed:
Implement recurring events

rycs2812 committed:
Merge branch 'branch-BCD-Extension'

rycs2812 committed:
Improve Duke GUI

The text in DialogBox would get truncated when too long, which might
prevent the full list of tasks from being displayed when "list" is
called.

This commit resolves this issue.

rycs2812 committed:
Polish Duke GUI

Added different visual formats for the user and duke dialog boxes,
changed padding and background colour of dialog box, and implemented
visual highlighting of command errors to the user.

rycs2812 committed:
Merge branch 'branch-A-BetterGui'

rycs2812 committed:
Enable vertical resizing of content as Window changes size

rycs2812 committed:
Add Ui.png

rycs2812 committed:
Update docs\README.md

rycs2812 committed:
Fix bug for 'event' command and change build.gradle mainClassName

Bug occurs when a valid recurring interval but invalid event date is
supplied.

rycs2812 committed:
Enable scrolling with mouse wheel on ScrollPane
